apps/consulting/posts/improvements-to-bsr-matrix-multiplication-in-pytorch.md

```diff
@@ -0,0 +1,292 @@
+optimal. However, as cuSPARSE did not support these data types a working implementation was created
+by composing calls into the dense kernel for each sparse block. This approach suffers from a
+significant penalty due to launching many CUDA kernels. 
+
+<img src="posts/improvements-to-bsr-matrix-multiplication-in-pytorch/bsr_dense_mm_half_vs_sparse.png"
+    name="fig-6"
+    alt="A plot displaying sparsity ratio (as a percentage) on the horizontal axis, and speedup over
+    the older sparse implementation on the vertical.  Three lines are plotted indicating block sizes
+    used for the sparse tensor of 16, 32, and 64."
+/>
+
+Comparing performance for half precision to the [dense baseline][#fig-7] we again see one clear win.
+Using a block size of 64 the triton implementation begins to out perform the dense matrix multiply
+at 60% sparsity.  Unfortunately, this does not hold for smaller block sizes with 32 reach >1x
+speedup at 90% sparsity, and block size 16 only passing this point at 99%. 
+
+<img src="posts/improvements-to-bsr-matrix-multiplication-in-pytorch/bsr_dense_mm_half_vs_dense.png"
+    name="fig-7"
+    alt="A plot displaying sparsity ratio (as a percentage) on the horizontal axis, and speedup over
+    the older sparse implementation on the vertical.  There is a horizontal line marking the speedup
+    ratio of 1.0 above which the sparse subject takes less time to execute than the dense
+    counterpart.  Three lines are plotted indicating block sizes used for the sparse tensor of 16,
+    32, and 64."
+/>
+
+## Try it out!
+
+These kernels will be shipped with PyTorch 2.1. They can all be found in the
+`torch.sparse._triton_ops` private submodule. Although not yet stable, the curious user may want
+experiment with these tools. If you want to stick to the public features only, the `bsr_dense_mm`
+kernel is also fully integrated with matrix multiply ops like `torch.mm` and  `torch.addmm`. 
+If you meet the following conditions:
+
+- The RHS of the matrix multiply or `A` in `A@B`, `torch.mm(A, B)`, or `torch.addmm(C, A, B)` has
+  BSR layout, and other tensors are dense.
+- The device type is `cuda`
+- The data type is `torch.float16`, or `torch.bfloat16`
+
+then the work will be forwarded to the triton kernel! It was an interesting process to set
+those bindings up, but that would deserve its own blog post. 
+
+## Conclusion
+
+We have made some serious improvements to the usability of key features for BSR, but we are not
+quite finished yet. These are only the first steps toward a larger goal, namely, making block sparse
+layouts a first class technique for accelerating transformer workflows. With PyTorch 2.1 coming out
+it is a good time to take stock of where we are at, and we have come a long way, but there is plenty
+of work left to do. In the coming months we hope to improve on these kernels and add more. We also
+are working to enable pathways such that masking behavior for gradients can be accessed without
```

Comment:
"We also are" or "We are also"?

Reply:
I think both ways are acceptable, but "We are also" reads better  will change it